File: myau2Extension/PluginApp/Parameters.swift
import AudioToolbox
import Foundation

let PB: ParameterSpecBuilder<ParameterAddress> = ParameterSpecBuilder()

let oscWaveValues = ["Saw", "Rect", "Tri", "Sine", "Noise"]

let lfoDestinationValues = [
  "None", "Osc1Pitch", "Osc1PWMix", "Osc1Volume", "Osc2Pitch", "Osc2PWMix", "Osc2Volume",
  "FilterCutoff", "AmpVolume",
]

struct DefaultValues {
  // static let osc1Wave = "Saw"
  // static let osc1Volume: Float = 1
  //debug
  static let osc1Wave = "Sine"
  static let osc1Volume: Float = 0.1
}

let myPluginParameterSpecs = ParameterTreeSpec {
  ParameterGroupSpec(identifier: "global", name: "Global") {
    PB.Bool(.osc1On, "osc1On", "OSC1 On", true)
    PB.Enum(.osc1Wave, "osc1Wave", "OSC1 Wave", DefaultValues.osc1Wave, oscWaveValues)
    PB.Unary(.osc1Octave, "osc1Octave", "OSC1 Octave", 0.5)
    PB.Unary(.osc1PwMix, "osc1PwMix", "OSC1 PwMix", 0.5)
    PB.Unary(.osc1Volume, "osc1Volume", "OSC1 Volume", DefaultValues.osc1Volume)
    //
    PB.Bool(.osc2On, "osc2On", "OSC2 On", false)
    PB.Enum(.osc2Wave, "osc2Wave", "OSC2 Wave", "Saw", oscWaveValues)
    PB.Unary(.osc2Octave, "osc2Octave", "OSC2 Octave", 0.5)
    PB.Unary(.osc2Detune, "osc2Detune", "OSC2 Detune", 0.5)
    PB.Unary(.osc2Volume, "osc2Volume", "OSC2 Volume", 0.5)
    //
    PB.Bool(.filterOn, "filterOn", "Filter On", true)
    PB.Enum(.filterType, "filterType", "Filter Type", "LPF", ["LPF", "BPF", "HPF"])
    PB.Unary(.filterCutoff, "filterCutoff", "Filter Cutoff", 1.0)
    PB.Unary(.filterPeak, "filterPeak", "Filter Peak", 0.0)
    PB.Unary(.filterEnvMod, "filterEnvMod", "Filter EnvMod", 0.5)
    //
    PB.Bool(.ampOn, "ampOn", "Amp On", true)
    PB.Unary(.ampAttack, "ampAttack", "Amp Attack", 0.0)
    PB.Unary(.ampDecay, "ampDecay", "Amp Decay", 0.0)
    PB.Unary(.ampSustain, "ampSustain", "Amp Sustain", 1.0)
    PB.Unary(.ampRelease, "ampRelease", "Amp Release", 0.0)
    //
    PB.Bool(.lfoOn, "lfoOn", "LFO On", false)
    PB.Enum(.lfoWave, "lfoWave", "LFO Wave", "Saw", oscWaveValues)
    PB.Unary(.lfoRate, "lfoRate", "LFO Rate", 0.5)
    PB.Unary(.lfoDepth, "lfoDepth", "LFO Depth", 0.5)
    PB.Enum(.lfoTarget, "lfoTarget", "LFO Target", "None", lfoDestinationValues)
    //
    PB.Bool(.egOn, "egOn", "EG On", false)
    PB.Unary(.egAttack, "egAttack", "EG Attack", 0.0)
    PB.Unary(.egDecay, "egDecay", "EG Decay", 0.0)
    PB.Unary(.egAmount, "egAmount", "EG Amount", 0.5)
    PB.Enum(.egTarget, "egTarget", "EG Target", "None", lfoDestinationValues)
    //
    PB.Unary(.glide, "glide", "glide", 0.0)
    PB.Enum(.voicingMode, "voicingMode", "Voicing Mode", "Mono", ["Mono", "Poly"])
    PB.Unary(.masterVolume, "masterVolume", "Master Volume", 0.8)
  }
}

File: v1-xcode/dsp-dev/definitions/parameters.ts
export enum OscWave {
  Saw = 0,
  Rect,
  Tri,
  Sine,
  Noise,
}
export enum LfoWave {
  Sine = 0,
  Tri,
  Rect,
  Saw,
  SampleHold,
}

export enum FilterType {
  LPF = 0,
  BPF,
  HPF,
}

export enum ModTarget {
  None = 0,
  Osc1Pitch,
  Osc1PwMix,
  Osc1Volume,
  Osc2Pitch,
  Osc2Volume,
  FilterCutoff,
  AmpVolume,
}

export enum VoicingMode {
  Poly = 0,
  Mono,
}

export type SynthParametersSuit = {
  osc1On: boolean;
  osc1Wave: OscWave;
  osc1Octave: number;
  osc1PwMix: number;
  osc1Volume: number;
  //
  osc2On: boolean;
  osc2Wave: OscWave;
  osc2Octave: number;
  osc2Detune: number;
  osc2Volume: number;
  //
  filterOn: boolean;
  filterType: FilterType;
  filterCutoff: number;
  filterPeak: number;
  filterEnvMod: number;
  //
  ampOn: boolean;
  ampAttack: number;
  ampDecay: number;
  ampSustain: number;
  ampRelease: number;
  //
  lfoOn: boolean;
  lfoWave: LfoWave;
  lfoRate: number;
  lfoDepth: number;
  lfoTarget: ModTarget;
  //
  egOn: boolean;
  egAttack: number;
  egDecay: number;
  egAmount: number;
  egTarget: ModTarget;
  //
  glide: number;
  voicingMode: VoicingMode;
  masterVolume: number;
};

export const defaultSynthParameters: SynthParametersSuit = {
  osc1On: true,
  osc1Wave: OscWave.Saw,
  osc1Octave: 0.5,
  osc1PwMix: 0,
  osc1Volume: 1,
  //
  osc2On: false,
  osc2Wave: OscWave.Saw,
  osc2Octave: 0.5,
  osc2Detune: 0.5,
  osc2Volume: 1,
  //
  filterOn: true,
  filterType: FilterType.LPF,
  filterCutoff: 1,
  filterPeak: 0,
  filterEnvMod: 0,
  //
  ampOn: true,
  ampAttack: 0,
  ampDecay: 0,
  ampSustain: 1,
  ampRelease: 0,
  //
  lfoOn: false,
  lfoWave: LfoWave.Sine,
  lfoRate: 0.5,
  lfoDepth: 0.5,
  lfoTarget: ModTarget.Osc1Pitch,
  //
  egOn: false,
  egAttack: 0,
  egDecay: 0.5,
  egAmount: 0.5,
  egTarget: ModTarget.FilterCutoff,
  //
  glide: 0,
  voicingMode: VoicingMode.Poly,
  masterVolume: 0.5,
};
